Post message

Sly-Agustin opened the pull request:
Dejo hecho el pull request con la implementación de la request post.

Sly-Agustin committed:
Empezo la tarea de DB

- Agregados módulos de mysql
- Establecida la conexión con la db local
- Agregadas querys de prueba

Sly-Agustin committed:
Update package.json

- Agregado el script para iniciar la consulta a la base de datos

Sly-Agustin committed:
Modelos y seeders

- Agregado seeder y modelos

Sly-Agustin committed:
Agregado otro modelo

Sly-Agustin committed:
Delete node_modules directory

Sly-Agustin committed:
Agregado gitignore

- Agregado gitignore, los módulos hbra que instalarlos con "npm install" cuando se pullee.
- Update al readme

Sly-Agustin committed:
Merge branch 'main' of https://github.com/Sly-Agustin/neocomplexx

Sly-Agustin committed:
Corregida ruta

- Corregida una ruta con nombre incorrecto que provocaba que no se pudiera iniciar el script start

Sly-Agustin committed:
Borrado codigo

- Borrado codigo innecesario

Sly-Agustin committed:
Update al readme, paquetes y seeder

- Update al readme para reflejar los últimos cambios
- Agregados paquetes al package.json.
- Creado el docker-compose para inicializar la base de datos, esta base de datos es temporal y una vez detenido el servicio se borra.
- Fixeado un typo que hacía que el seeder no funcionara

Sly-Agustin committed:
Implemented the post request

- Added a post request to add employees
- Added a router for employees
- Updated readme with steps to use the new route

Sly-Agustin committed:
Update README.md

- Corrections to readme

Sly-Agustin committed:
Update README.md

Sly-Agustin committed:
Prettier

- Formateados los archivos según la configuracion de prettier

Sly-Agustin committed:
Cambios del code review

- Cambios a las migraciones
- Cambios al package.json y agregados modulos de nodemon
- Agregado un controlador para el empleado

Sly-Agustin committed:
Manejo de errores

- Actualizadas las validaciones
- Ahora se detecta si algún atributo no es del tipo que se requiere
- Agregado try-catch a la hora de insertar en la base de datos,  se capturan errores no contemplados y se devuelve un status 400.

Sly-Agustin committed:
Agregadas rutas

- Agregada una ruta para buscar todos los empleados
- Agregada una ruta para buscar un empleado por  ID
- Modificado el controlador de employee
- Pasada de prettier a todos los archivos (por eso se modifico mucho en el commit)

Sly-Agustin committed:
Update employeesController.js

- Arreglado un crash cuando se buscaba un error.
- La consulta por un empleado ahora devuelve solo el nombre, apellido y cargo en lugar de todo lo que hay en la base de datos

diazmaira commented on README.md:
```diff
@@ -1,17 +1,47 @@
+To use access your local database:
+
+-   Use the env.sample file as template, create a .env file and fill with your database credentials
+-   ` $ npm run localDatabase` will show a successful database message if credentials were correct
```
You can use a more appropiated name which describe the intention, for example, `checkDatabaseConnection`.

diazmaira commented on README.md:
```diff
@@ -1,17 +1,47 @@
+To use the add an employee to the database with a POST request:
+
+-   Be sure to have the docker and database up
+-   Use a program like Postman to make requests
```
Please, export and upload in the root path the Postman collection associated. You can take a look to this link: https://apitransform.com/how-to-export-postman-collection/

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,94 @@
+
+	try {
+		let aux = await db.empleados.create({
+			nombre: req.body.name,
```
Use "Destructuring" to access variables, instead of accessing every time to `req.body`.

Take a look at this link: https://github.com/ngminhtrung/You-Dont-Know-JS/blob/master/es6%20%26%20beyond/ch2.md#destructuring

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,94 @@
+			nombre: req.body.name,
+			apellido: req.body.lastName,
+			cargo: req.body.position,
+			createdAt: new Date(),
```
Everytime you create an `empleado`, both `createdAt` and `updatedAt` contain always the same value. Can you set by default `createdAt`and  `updatedAt` so you don't need to set up each time you create an `empleado`?
Take a look to it https://sequelize.org/v5/manual/models-definition.html

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,94 @@
+			message: 'Field content cant be a number, they must be characters',
+		});
+		return;
+	}
```
The function `newEmployee` contains 48 lines of code. It turns difficult to read.  This is a **bad smell** called _long method/function_.  Please, extract from line 5 to 25 and define a function called `validateEmployee`, which takes and object as parameter: `{ name, lastName, position}`.

Read about Long method bad smell https://refactoring.guru/smells/long-method

Sly-Agustin committed:
Parte de los cambios pedidos en el pull request

- Cambiado el nombre de "npm run localDatabase" a "npm run checkDatabaseConnection" en caso de que las credenciales en el .env sean correctos establecerá la conexión.
- Modificada la request health, ahora ya no devuelve el status en el cuerpo como lo hacía antes
- Agregado el archivo JSON con las requests para importar en postman
- Actualizado el readme para mayor claridad y con los cambios a todo lo anterior

Sly-Agustin committed:
Update employeesController.js

- Destructurado el acceso a las variables mediante un método "getAttributesFromRequest"

Sly-Agustin committed:
Modificación al modelo de empleados

- Modificado el modelo de "empleados" de forma que la fecha de creación y modificación sea seteada a la actual por defecto sin necesidad de hacerlo en el controlador
- Eliminado la creación y el update de la fecha del employeesController

Sly-Agustin committed:
Clean code

- Creados métodos para modularizar mejor y achicar las funciones
- Modificada la export de postman con una request más

Sly-Agustin committed:
Update employeesController.js

- Borrado un console log que quedó suelto

Sly-Agustin committed:
Agregada una migracion

- Agregada una migración para remover la columna DNI.
- Actualizado el gitignore

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,142 @@
+}
+
+function attributesInRequestDefined(req, res) {
+	let {name: name, lastName: lastName, position: position} = getAttributesFromRequest(req);
```
If you don't need to re-name variables, you can use `object desttructuring`as follow:

`let {name, lastName, position} = ... `

Sly-Agustin replied:
Resolved in commit 83d7de4ed11bca24c689d1fe577bc38c8e589539

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,142 @@
+}
+
+function attributesInRequestDefined(req, res) {
+	let {name: name, lastName: lastName, position: position} = getAttributesFromRequest(req);
```
No need to define a function to get the attributes. You can use `desttrucuring object` as follow:
`const { name, lastName, position } = req.body;`

Sly-Agustin replied:
Resolved in 83d7de4ed11bca24c689d1fe577bc38c8e589539

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,142 @@
+}
+
+function checkTypeOfAttributes(req, res) {
+	let {name: name, lastName: lastName, position: position} = getAttributesFromRequest(req);
```
Same as line 29

Sly-Agustin replied:
Resolved in 83d7de4ed11bca24c689d1fe577bc38c8e589539

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,142 @@
+		return;
+	}
+
+	let {name: name, lastName: lastName, position: position} = getAttributesFromRequest(req);
```
Same as line 29

Sly-Agustin replied:
Resolved in 83d7de4ed11bca24c689d1fe577bc38c8e589539

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,142 @@
+	}
+}
+function validateEmployee(req, res) {
+	if (attributesInRequest(req, res) && attributesInRequestDefined(req, res) && checkTypeOfAttributes(req, res)) {
```
No need to use  `if` statement. Use functions as conditions as follow:
`return attributesInRequest(req, res) && attributesInRequestDefined(req, res) && checkTypeOfAttributes(req, res)`

Sly-Agustin replied:
Resolved in 83d7de4ed11bca24c689d1fe577bc38c8e589539

Sly-Agustin committed:
Destructuring

- Cambiado un método para que haga uso del "destructuring" en lugar de la asignación clasica

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,142 @@
+				errorMessage: err.parent.sqlMessage,
+			});
+		}
+		res.status(400).send({
```
This error could be defined as `500`, because it is a server error (database connection problems for example). 
**Server error codes** indicates that the request was accepted, but that an error on the server prevented the fulfillment of the request.

Sly-Agustin replied:
Resolved in 83d7de4ed11bca24c689d1fe577bc38c8e589539

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,142 @@
+}
+
+async function allEmployees(req, res) {
+	let employees = await db.empleados.findAll({ attributes: ['id', 'nombre', 'apellido'] });
```
use `try ...catch` in case accessing to database throws an error. Return `500` status code in the `catch` block

Sly-Agustin replied:
Resolved in 83d7de4ed11bca24c689d1fe577bc38c8e589539

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,142 @@
+}
+
+async function getEmployeeById(req, res) {
+	let id = parseInt(req.params.id, 10);
```
`req.params.id` is accessed more than once. So, it is useful to define a variable, for example, `const id = req.params.id`.

Sly-Agustin replied:
Resolved in 113562c8629a8d273dc780bec00a4a2de01fa91a

diazmaira commented on controllers/employeesController.js:
```diff
@@ -0,0 +1,142 @@
+			employee: employee[0],
+		});
+	} catch (err) {
+		res.status(400).send({
```
Same as line `72`.

Sly-Agustin replied:
Resolved in 83d7de4ed11bca24c689d1fe577bc38c8e589539

Sly-Agustin committed:
Resolves del PR

Resueltos:
- https://github.com/Sly-Agustin/neocomplexx/pull/1#discussion_r776814609
- https://github.com/Sly-Agustin/neocomplexx/pull/1#discussion_r776815055
- https://github.com/Sly-Agustin/neocomplexx/pull/1#discussion_r776815205
- https://github.com/Sly-Agustin/neocomplexx/pull/1#discussion_r776815285
- https://github.com/Sly-Agustin/neocomplexx/pull/1#discussion_r776815798
- https://github.com/Sly-Agustin/neocomplexx/pull/1#discussion_r776816927
- https://github.com/Sly-Agustin/neocomplexx/pull/1#discussion_r776817651
- https://github.com/Sly-Agustin/neocomplexx/pull/1#discussion_r776818488

Sly-Agustin committed:
Arreglado issue

- Arreglado issue https://github.com/Sly-Agustin/neocomplexx/pull/1#discussion_r776818235
